Feature dags projects

jdddog opened the pull request:
This pull request adds the ability to run third party projects that contain DAGs within the Observatory Platform. To add a third party DAG project you need to customise a part of the config.yaml file:

```yaml
# User defined Observatory DAGs projects:
dags_projects:
  - package_name: observatory-dags  # the third party DAGs package name
    path: /home/user/observatory-platform/observatory-dags  # the path to the package on the local file system
    dags_module: observatory.dags.dags  # the Python module import path to the module that contains the DAGs
```

The observatory has been split into three different namespace packages, so that each package can be installed independently:
* observatory-platform: the command line interface, common utility functions 
* observatory-dags: the package that contains the main telescopes and workflows. This is loaded into the observatory platform in the same way as a third party DAGs package is.
* observatory-reports: the reports package.

Other changes:
* Made an iteration of the observatory configuration file format. This was to enable the ability to specify third party DAGs projects as dependencies and to enable most properties in the configuration file to be optional, especially when the observatory is run locally.
* Command line interface code tidied up, moving core functionality into classes that represent the individual commands being executed.
* Added the ability to customise the following options from the observatory platform command line interface: redis port, flower ui port, airflow ui port, elastic port, kibana port, external docker network name and the docker compose container namespace name.

I'm working on getting these changes functioning in Terraform, which will be added to this pull request later.

In the mean time it would be good to get everyone's feedback on these changes.

tuanchien commented:
Good work Jamie.  Seems alright to me. One thing I'm unsure about though is the cd <project> && pip install -e .
Perhaps might be better to do pip install -e <project> instead
The 'cd' bit will finish in that project directory, and someone less experienced with the command line might be confused when the next commands fail because you're no longer in the right directory to execute the next commands.

aroelo reviewed (commented):
Hey Jamie, great work overall.
It's nice to see everything become better structured and organised, bit by bit, especially considering the config file as well.

Curious to see the coming changes regarding the terraform configuration.

Some questions/remarks:
*  Due to some changes the DEBUG_FILE_PATH property is commented out in most telescopes. This is used in the 'develop' environment. I know we want to change this approach anyway, but the telescopes now don't run in the 'develop' environment, since it can't find the debug file path.

* At the moment it is not possible to add or delete variables to the terraform cloud workspace (during update-workspace), because of the schema and how new variables are detected. Would it make sense to simplify code by removing everything that assumes variables will be added/deleted? Or should we keep this in for possible use in the future?

* I noticed that the config-terraform.yml template includes the 'dags_projects' section, but this is not read in as a terraform variable. Could it be left out? Or will it be included with some of the terraform changes?

* As mentioned on slack, I'm having some issues with an import error from the click module, so like you said maybe renaming the `observatory-platform/observatory/platform/cli/click.py` file would be a good idea to prevent this.

cameronneylon commented on observatory-reports/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+observatory-platform
+matplotlib==3.1.*
+pandas==0.25.*
+seaborn==0.10.*
```
Should this be 0.11.* ?

jdddog replied:
Thanks Cameron, yeah you're right, I've changed it to that version now.

cameronneylon reviewed (commented):
One broad question is the extent to which reports should depend on platform? Is there actually a close dependency in practice? The main link would be the structure of tables and how reports could act as an example of abstracting queries and structure away from the user. 

Otherwise all makes sense and seems good. One minor issue with the seaborn version but that's not a big issue as we'll be moving away from that anyway.

jdddog commented:
> Good work Jamie. Seems alright to me. One thing I'm unsure about though is the cd && pip install -e .
> Perhaps might be better to do pip install -e instead
> The 'cd' bit will finish in that project directory, and someone less experienced with the command line might be confused when the next commands fail because you're no longer in the right directory to execute the next commands.

Thanks Tuan, I've updated the installation instructions with this change.

jdddog commented:
> One broad question is the extent to which reports should depend on platform? Is there actually a close dependency in practice? The main link would be the structure of tables and how reports could act as an example of abstracting queries and structure away from the user.
> 
> Otherwise all makes sense and seems good. One minor issue with the seaborn version but that's not a big issue as we'll be moving away from that anyway.

Thanks Cameron, that's a good point, I've managed the remove the dependency to the observatory-platform. There were a couple of functions from observatory-platform that were used in the unit tests for observatory-reports, which I've moved out of the observatory-platform.

jdddog commented:
> Hey Jamie, great work overall.
> It's nice to see everything become better structured and organised, bit by bit, especially considering the config file as well.
> 
> Curious to see the coming changes regarding the terraform configuration.

Thanks Aniek.

> 
> Some questions/remarks:
> 
> * Due to some changes the DEBUG_FILE_PATH property is commented out in most telescopes. This is used in the 'develop' environment. I know we want to change this approach anyway, but the telescopes now don't run in the 'develop' environment, since it can't find the debug file path.

Yeah you're right, I've started making some changes to address this, I just need to double check them.

> * At the moment it is not possible to add or delete variables to the terraform cloud workspace (during update-workspace), because of the schema and how new variables are detected. Would it make sense to simplify code by removing everything that assumes variables will be added/deleted? Or should we keep this in for possible use in the future?

I'm happy to leave it in case we need this functionality in the future.

> * I noticed that the config-terraform.yml template includes the 'dags_projects' section, but this is not read in as a terraform variable. Could it be left out? Or will it be included with some of the terraform changes?

The `dags_projects` section isn't set as a Terraform variable, however it is used as part of the process for deploying the system with Terraform. It is used by the DAG load_dags.py to determine what DAG directories to load in Airflow, by Docker to determine what volumes need to be mounted and the Packer image builder to determine what packages to include in the build.

> * As mentioned on slack, I'm having some issues with an import error from the click module, so like you said maybe renaming the `observatory-platform/observatory/platform/cli/click.py` file would be a good idea to prevent this.

I've fixed this now by renaming it to click_utils.py and the tests as test_click_utils.py.

jdddog commented on observatory-platform/observatory/platform/docker/entrypoint-airflow.sh.jinja2:
```diff
@@ -11,21 +27,38 @@ if [ $1="webserver" ]; then
 TESTS_DIRECTORY='/home/airflow/.local/lib/python3.7/site-packages/tests'
 if [ -d ${TESTS_DIRECTORY} ]; then
   rm -r ${TESTS_DIRECTORY}
 fi
```
Ah I must have forgotten about that. Once I've finished the changes for the DEBUG_FILE_PATH we should be able to remove it because we won't need to depend on the tests package anymore.

jdddog commented on observatory-platform/observatory/platform/config-terraform.yaml.jinja2:
```diff
@@ -0,0 +1,58 @@
+# airflow_connections:
+#   my_connection: http://my-username:my-password@
+
+# User defined Observatory DAGs projects:
+# dags_projects:
+#   - package_name: observatory-dags
+#     path: {{ observatory_dags_path }}
+#     dags_module: observatory.dags.dags
```
Yeah it will be used with the changes to Terraform

jdddog commented on docs/deploy_terraform.md:
````diff
@@ -17,7 +17,7 @@ sudo chmod +x /usr/local/bin/packer
 Install Terraform:
 ```bash
-sudo curl -L "https://releases.hashicorp.com/terraform/0.13.0-beta3/terraform_0.13.0-beta3_linux_amd64.zip" -o /usr/local/bin/terraform
+sudo curl -L "https://releases.hashicorp.com/terraform/0.13.4/terraform_0.13.4_linux_amd64.zip" -o /usr/local/bin/terraform
````
Thanks Aniek. Does 0.13.5 work on Mac? I'm using that version on Linux now.

aroelo replied:
Just checked 0.13.5 and works without issues :-). I noticed that the terraform cloud workspace versions now can also be updated.

If you could update the version to "0.13.5" here that would be great:
https://github.com/The-Academic-Observatory/observatory-platform/blob/2eaccf376805a3f7f59fce6b6bb9fb649bc66421/observatory-platform/observatory/platform/terraform_api.py#L152

I also made a separate organisation for unit testing a while back, but didn't use it for tests yet as it didn't have the version available that we needed for some functionalities. We could use that one now, or create another one with a different name if you'd prefer. (The one I made is called observatory-unit-testing). Just need to update the token in github secrets if we do.

jdddog replied:
Thanks Aniek.

Yeah it would be good to use the observatory-unit-testing Terraform organisation for tests, I added the an extra Github secret that you can supply the organisation name with: `TESTS_TERRAFORM_ORGANISATION`

jdddog commented on README.md:
```diff
@@ -132,31 +142,63 @@ config.yaml saved to: "/home/user/.observatory/config.yaml"
 Please customise the parameters with '<--' in the config file. Parameters with '#' are optional.
```
That's a good point. I'll remove that for the local config file because they are stilled used with the terraform config file.

jdddog commented on docs/observatory_dev.md:
```diff
@@ -16,31 +16,63 @@ config.yaml saved to: "/home/user/.observatory/config.yaml"
+| mag_releases_table     | `http://myname:mypassword@myhost.com`     | Stores the azure-storage-account-name as a login and url-encoded-sas-token as password     |
+| mag_snapshots_container     | `http://myname:mypassword@myhost.com`     | Stores the azure-storage-account-name as a login and url-encoded-sas-token as password     |
+| terraform     | `mysql://:terraform-token@`     | Stores the terraform user token as a password (used to create/destroy VMs)     |
+| slack     | `https://:T00000000%2FB00000000%2FXXXXXXXXXXXXXXXX XXXXXXXX@https%3A%2F%2Fhooks.slack.com%2Fservices`     | Stores the URL for the Slack webhook as a host and the token as a password     |
```
Thanks, I didn't notice that.

The sphinx_markdown_tables package looked good, however it has a GPL-3.0 license. I've fixed it by enabling RST to be evaluated in the markdown files, for example:
```
```eval_rst
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
| Connection Key          | Example                                                                                              | Description                                                                            |
+=========================+======================================================================================================+========================================================================================+
| crossref                | http://myname:mypassword@myhost.com                                                                  | Stores the token for the crossref API as a password                                    |
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
| mag_releases_table      | http://myname:mypassword@myhost.com                                                                  | Stores the azure-storage-account-name as a login and url-encoded-sas-token as password |
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
| mag_snapshots_container | http://myname:mypassword@myhost.com                                                                  | Stores the azure-storage-account-name as a login and url-encoded-sas-token as password |
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
| terraform               | mysql://:terraform-token@                                                                            |   Stores the terraform user token as a password (used to create/destroy VMs)           |
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
| slack                   | https://:T00000000%2FB00000000%2FXXXXXXXXXXXXXXXXXXXXXXXX@https%3A%2F%2Fhooks.slack.com%2Fservices   | Stores the URL for the Slack webhook as a host and the token as a password             |
+-------------------------+------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------+
```
```

jdddog commented on observatory-platform/observatory/platform/observatory_config.py:
```diff
@@ -0,0 +1,1023 @@
+                TerraformVariable('database', self.airflow.to_hcl(), sensitive=sensitive, hcl=True),
+                TerraformVariable('airflow_main_vm', self.airflow_main_vm.to_hcl(), hcl=True),
+                TerraformVariable('airflow_worker_vm', self.airflow_worker_vm.to_hcl(), hcl=True),
+                TerraformVariable('airflow_variables', list_to_hcl(self.airflow_variables), hcl=True,
```
Yeah that's a good point, I can do that.

It would probably be good to be able to turn sensitive off from the command line for debugging purposes for all of the Terraform variables, but I can do that later.

aroelo reviewed (commented):
Hi Jamie,

Nice work!
I've got a couple of questions/remarks (some quite small).
Because I already viewed some of the code earlier, I can't always comment in a file, sorry about that.

* The `startup.tpl.jinja2` file is missing a license/author

* I saw that there are two `is_environment_valid` properties for the terraform commands. One inside the `TerraformBuilder` and one inside the `TerraformCommand` class. The one inside the TerraformBuilder isn't used when you run build-terraform or build-image, so the dependency for packer is not checked.

* The debug property of the TerraformBuilder is hardcoded to True, would it be possible to add an option '--debug', like with the observatory platform commands?

* Related to this, while testing out the code I find it helps me to see the subprocess commands printed out to see what is actually going on. Perhaps it would be nice to print these for both terraform & platform commands. For example in the `build_image` method of the TerraformBuilder something like this, just an idea:
```
        if self.debug:
            print('Executing subprocess:')
            print(indent(f'Command: {subprocess.list2cmdline(args)}', INDENT1))
            print(indent(f'Cwd: {self.terraform_build_path}', INDENT1))
```

* If I understand correctly, updating the build files (with build-terraform), but not the VM image would only be useful for changes in the terraform configuration, e.g. running a terraform apply afterwards. For any changes in the config file like variables/connections you would have to rebuild the image as well(?) Perhaps it would be useful to have some documentation on in which cases you need to use `build-image` vs `build-terraform`.

* When working with multiple environments (develop/staging/production), I think all of these will build their files in the same directory (~/.observatory/build), should we maybe create separate directories for the different environments if you want to more easily switch between these?

*  I noticed there is some code inside the Dockerfile to install requirements from DAG projects in different ways, depending on the project type, but for now the project type is always set to 'local'. Do you plan to add this functionality in the future?

* When creating a new Terraform Cloud workspace today I noticed that I tried a different prefix name in the config, this will create a workspace just fine, but won't work actually trying to connect the workspace to the terraform configuration files. In the 'main.tf' file, the prefix is hardcoded to 'observatory-', so the workspace in Terraform Cloud needs to have the same workspace prefix. 
I think we might need to eliminate this option from the terraform configuration and hardcode the prefix name. This is my mistake, I think I was always just using 'observatory-' so it didn't give any issues so far. 

* A more general question, but why do we need to pull and build Docker containers both in 'build.sh' when creating the image as well as in the startup script of the VM? Wouldn't it be sufficient to just build the docker containers when creating the image and doing 'docker-compose up' in the startup script, or not build them yet in build.sh?

jdddog commented on docs/deploy_terraform.md:
```diff
@@ -110,8 +110,9 @@ Storage Transfer Admin
+google project. This is required for Packer to create the image. Other Google Cloud services are enabled by Terraform 
+itself.
 
 ## Clone and checkout files
```
That's a good point. I've changed this now and also added at the start of the tutorial that you should have installed the Observatory Platform and given a link to that tutorial.

jdddog commented on docs/deploy_terraform.md:
````diff
@@ -255,16 +352,18 @@ To destroy the system with Terraform:
-terraform destroy -target module.airflow_worker_vm.google_compute_instance.vm_instance -var-file=dev.tfvars terraform
+## Manually destroy the VMs
+```
+terraform apply -target module.airflow_main_vm.google_compute_instance.vm_instance
+terraform apply -target module.airflow_worker_vm.google_compute_instance.vm_instance
````
Thanks, I've updated this

jdddog commented:
> Hi Jamie,
> 
> Nice work!
> I've got a couple of questions/remarks (some quite small).
> Because I already viewed some of the code earlier, I can't always comment in a file, sorry about that.
> 

Thanks Aniek!

> * The `startup.tpl.jinja2` file is missing a license/author

Thanks, I missed that, I've added it in.

> * I saw that there are two `is_environment_valid` properties for the terraform commands. One inside the `TerraformBuilder` and one inside the `TerraformCommand` class. The one inside the TerraformBuilder isn't used when you run build-terraform or build-image, so the dependency for packer is not checked.
> * The debug property of the TerraformBuilder is hardcoded to True, would it be possible to add an option '--debug', like with the observatory platform commands?
> * Related to this, while testing out the code I find it helps me to see the subprocess commands printed out to see what is actually going on. Perhaps it would be nice to print these for both terraform & platform commands. For example in the `build_image` method of the TerraformBuilder something like this, just an idea:
> 
> ```
>         if self.debug:
>             print('Executing subprocess:')
>             print(indent(f'Command: {subprocess.list2cmdline(args)}', INDENT1))
>             print(indent(f'Cwd: {self.terraform_build_path}', INDENT1))
> ```
> 

Thanks, I've updated the project to address these.

> * If I understand correctly, updating the build files (with build-terraform), but not the VM image would only be useful for changes in the terraform configuration, e.g. running a terraform apply afterwards. For any changes in the config file like variables/connections you would have to rebuild the image as well(?) Perhaps it would be useful to have some documentation on in which cases you need to use `build-image` vs `build-terraform`.

Yeah it is a bit unclear, I've updated the documentation on the Read the Docs tutorial, a summary is below.

For the `build-image` command, use this command if you have:
* Created, removed or updated user defined Observatory DAGs projects via the field `dags_projects`, in the Observatory
Terraform config file.
* Updated any code in the Observatory Platform.
* Update the `backend.environment` variable in the Observatory Terraform config file: you need to make sure that an
image is built for the other environment.

You will need to taint the VMs and update them so that they use the new image.

You do not need to run the `build-image` command if:
* You have created, removed or updated user defined Apache Airflow connections or variables in the Observatory
Terraform config file: in this case you will need to update the Terraform workspace.
* You have changed any other settings in the Observatory Terraform config file (apart from `backend.environment`): 
in this case you will need to update the Terraform workspace variables and run `terraform apply`.

For the `build-terraform` command:
* You only need to run it if you have updated the Terraform deployment scripts, but nothing else.

> * When working with multiple environments (develop/staging/production), I think all of these will build their files in the same directory (~/.observatory/build), should we maybe create separate directories for the different environments if you want to more easily switch between these?

Yeah I thought about this as well. The main thing against it is that you can also have multiple branches for each environment, so if you were going to get the full benefit then you would need to have a different folder for each environment and branch. It doesn't take long to generate these files, it is more the inconvience of having to run the commands to generate them.

Development tools usually just build files into a single directory, but in those cases you don't need to run multiple commands to build the project.

This wouldn't be a problem if we go with your idea of wrapping up the Terraform commands into our own command line interface, because then we can run all of the build steps, including building the Terraform files and building the Google Compute VM image (if certain settings or files have changed) before running the Terraform commands. It would also mean that we can get rid of the `build-image` and `build-terraform` commands as people won't need to use them. I would prefer to leave that task for another day as we need to think through all of the implications.

What do you think?

> * I noticed there is some code inside the Dockerfile to install requirements from DAG projects in different ways, depending on the project type, but for now the project type is always set to 'local'. Do you plan to add this functionality in the future?

Yes, we can add that functionality in a separate release, because we don't need it right away.

> * When creating a new Terraform Cloud workspace today I noticed that I tried a different prefix name in the config, this will create a workspace just fine, but won't work actually trying to connect the workspace to the terraform configuration files. In the 'main.tf' file, the prefix is hardcoded to 'observatory-', so the workspace in Terraform Cloud needs to have the same workspace prefix.
>   I think we might need to eliminate this option from the terraform configuration and hardcode the prefix name. This is my mistake, I think I was always just using 'observatory-' so it didn't give any issues so far.

I've removed the workspace prefix as an option and hardcoded it.

> * A more general question, but why do we need to pull and build Docker containers both in 'build.sh' when creating the image as well as in the startup script of the VM? Wouldn't it be sufficient to just build the docker containers when creating the image and doing 'docker-compose up' in the startup script, or not build them yet in build.sh?

Just in case the Docker images have changed since the VM image was built. If they haven't changed then the pull and build commands won't do anything. It might be unnecessary...

tuanchien commented on observatory-platform/observatory/platform/terraform_api.py:
```diff
@@ -109,82 +209,83 @@ def workspace_id(self, organisation: str, workspace: str) -> str:
-        """
-        Add a new variable to the workspace. Will return an error if the variable already exists.
-        :param attributes: attributes of the variable
+    def add_workspace_variable(self, variable: TerraformVariable, workspace_id: str) -> str:
```
For clarity, in the functions that follow, maybe make named variables for the http response codes instead of referring to their numeric value

jdddog replied:
I've replaced these with HTTPStatus code enums.

tuanchien reviewed (changes requested):
Looking good Jamie.  I've left a few suggestions.

jdddog commented on observatory-platform/observatory/platform/cli/terraform_command.py:
```diff
@@ -0,0 +1,193 @@
+        """
+
+        if self.debug:
+            return 2
```
That's a good point, I've removed the magic numbers.

jdddog commented on observatory-platform/observatory/platform/docker/entrypoint-airflow.sh.jinja2:
```diff
@@ -0,0 +1,61 @@
+cd /opt/observatory-platform
+
+# Set PBR version for installing packages
+export PBR_VERSION=0.0.1
```
The version isn't really important for this, it just matters that it installs.

jdddog commented on observatory-platform/observatory/platform/cli/cli.py:
```diff
@@ -0,0 +1,432 @@
+              is_flag=True,
+              default=DEBUG,
+              help='Print debugging information.')
+def platform(command: str, config_path: str, build_path: str, dags_path: str, data_path: str, logs_path: str,
```
Thanks, I've split these up more now.

jdddog commented on observatory-platform/observatory/platform/cli/cli.py:
```diff
@@ -0,0 +1,432 @@
+              is_flag=True,
+              default=DEBUG,
+              help='Print debugging information.')
+def terraform(command, config_path, terraform_credentials_path, debug):
```
Thanks, I've split these up more now.
